File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        def recursion(node, min_, max_):
            if node is None:
                return 0
            
            current = max(node.val - min_, max_ - node.val)
            min_ = min(min_, node.val)
            max_ = max(max_, node.val)

            return max(
                current,
                recursion(node.left, min_, max_),
                recursion(node.right, min_, max_),
            )
        
        return recursion(root, root.val, root.val)

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode* root,int m,int M){
        if(root==NULL)return 0;
        if(m==-1)m = root->val;
        if(M==-1)M = root->val;
        if(root->left==NULL && root->right==NULL){
            int x = m,X = M;
            if(x>root->val)x = root->val;
            if(X<root->val)X = root->val; 
            return X-x;
        }
        if(m>root->val){
            m = root->val;
        }
        if(M<root->val){
            M = root->val;
        }
        return max(solve(root->left,m,M),solve(root->right,m,M));
    }
    int maxAncestorDiff(TreeNode* root) {
        return solve(root,-1,-1);
    }
};